### DISCUSSION TITLE
Multiple param values in fixture

### DISCUSSION THREAD

**Discussion by DavideCanton** [pytest-dev/pytest#10899] [Q&A]
Hi all, I've noticed a not very well documented behavior in parametrizing fixtures.

A fixture with params errors out in the discovery if the params have multiple values. I believe it's because a single name fixture cannot bind different parameter values, but I didn't find anything about it in the docs, neither the error is clear:

```py
import pytest


@pytest.fixture(params=[pytest.param(i, 5) for i in range(5)])
def my_fixture(request):
    return request.param


def test_foo(my_fixture):
    assert my_fixture in {1, 2, 3, 4, 5}
```

```
$ pytest tests/foo_test.py

# removed output for brevity

tests/foo_test.py::test_foo: in "parametrize" the number of names (1):
  ['my_fixture']
must be equal to the number of values (2):
  (0, 5)

# removed output for brevity
```

**Comment by portyu9** [comment]
Hi @DavideCanton 

`pytest.param(i, 5)` means **two parameter values**, but your fixture has only one parameter name (`my_fixture`). That is why pytest reports:

```text
number of names (1) ... must be equal to the number of values (2)
```

If you want `(i, 5)` to be the **single value** of `request.param`, wrap it as one object:

```python
@pytest.fixture(params=[pytest.param((i, 5)) for i in range(5)])
def my_fixture(request):
    return request.param
```

You can also omit `pytest.param()` entirely here:

```python
@pytest.fixture(params=[(i, 5) for i in range(5)])
def my_fixture(request):
    return request.param
```

Then:

```python
def test_foo(my_fixture):
    i, value = my_fixture
```

If the `5` was intended to be a test ID rather than another value, use the keyword argument instead:

```python
pytest.param(i, id="5")
```

Please mark this answer as accepted if it helped, thank you.

**Reply by DavideCanton** [reply]
I know how does `pytest.fixture` work, what I mean is that there isn't (or at least there wasn't at the time I opened the question, I didn't recheck) mention in the documentation that you cannot use multiple value parameters in a fixture.

**Reply by portyu9** [reply]
Thanks for the additional context. A parametrized fixture only has **one parameter slot**: the fixture itself. Each entry in `params` becomes the single value exposed as `request.param`.

So this does not work:

```python
@pytest.fixture(params=[pytest.param(i, 5) for i in range(5)])
```

because `pytest.param(i, 5)` creates a parameter set containing **two values**, while the fixture parametrization expects one.

If the intended fixture value is the pair `(i, 5)`, make that pair the single parameter:

```python
@pytest.fixture(params=[pytest.param((i, 5)) for i in range(5)])
def my_fixture(request):
    return request.param
```

or simply:

```python
@pytest.fixture(params=[(i, 5) for i in range(5)])
def my_fixture(request):
    return request.param
```

That said, I agree the documentation is not very explicit about this distinction. `pytest.param()` is documented as usable with parametrized fixtures, but it does not clearly state that a fixture accepts only one value per parameter set, unlike `@pytest.mark.parametrize()` where multiple argument names can consume multiple values.

Please mark this answer as accepted if it helped, thank you.
